File: SortingAlgorithms/C++/SortingAlgorithms/SortingAlgorithms/MergeSort.cpp
#include <iostream>
#include <fstream>

using namespace std;

void mergeSort(double *unsortedArray, int left, int right);
void merge(double *sortedArray, int left, int middle, int right);
void swap(double *num1, double *num2);
void printArray(double *array, int size);


int main() {
	fstream dataInput;

	// create a stream to read file
	dataInput.open("unsortedFile.txt", ios::in);
	
	// size of the array
	int size; 
	dataInput >> size;
	double *unsortedArray = new double[size];

	// after read size, it's neccessary to ignore the rest of that line to jump into next line
	dataInput.ignore(1000, '\n');

	// contain the entire line
	char takenString[255]; 
	dataInput.getline(takenString, 1000, '\n');
	
	// default, don't care because it is not  meaningful
	char* context = NULL;
	char * token; // retain each element (token) that extracted from the taken string

	// get the first element delimited by space character (" ")
	token = strtok_s(takenString, " ", &context); 
	for (int index = 0; index < size; index++) {
		unsortedArray[index] = atof(token);
		token = strtok_s(NULL, " ", &context); // continue to get next element delimited by space character (' ')
	}

	/*
	// Way 2 : It is an simple way to read data from the file (not general)
	dataInput >> size;
	double *unsortedArray = new double[size];
	for (int index = 0; index < size; index++) {
	dataInput >> unsortedArray[index];
	}
	*/
	dataInput.close();
	
	// display the original array and the sorted array
	cout << "Original Array : " << endl;
	printArray(unsortedArray, size);
	cout << "\n\nSorted Array (accending) : " << endl;
	mergeSort(unsortedArray, 0, size - 1);
	printArray(unsortedArray, size);

	getchar();
	return 0;
}

// Recursive algorithm for Merge Sort
void mergeSort(double *unsortedArray, int left, int right) {
	if (left < right) {
		int middle = (left + right) / 2;
		mergeSort(unsortedArray, left, middle);
		mergeSort(unsortedArray, middle + 1, right);
		merge(unsortedArray, left, middle, right);
	}
}

// Merge two sorted arrays into an root array
void merge(double *sortedArray, int left, int middle, int right) {
	int leftArraySize = middle - left + 1;
	int rightArraySize = right - middle;

	// dynamic memory allocation for two new arrays supporting merging progress
	double *leftArray = new double[leftArraySize];
	double *rightArray = new double[rightArraySize];
	
	for (int index = 0; index < leftArraySize; index++) {
		leftArray[index] = sortedArray[left + index];
	}

	for (int index = 0; index < rightArraySize; index++) {
		rightArray[index] = sortedArray[middle + 1 + index];
	}
	// after comparing each pair of elements, choose a smaller element and place it into proper position in the main array
	int leftArrayIndex = 0;
	int rightArrayIndex = 0;
	int mainArrayIndex = left;
	while ( (leftArrayIndex < leftArraySize) && (rightArrayIndex < rightArraySize) ) {
		if (leftArray[leftArrayIndex] < rightArray[rightArrayIndex]) {
			sortedArray[mainArrayIndex] = leftArray[leftArrayIndex];
			leftArrayIndex++;
		}
		else {
			sortedArray[mainArrayIndex] = rightArray[rightArrayIndex];
			rightArrayIndex++;
		}
		mainArrayIndex++;
	}
	// add remaining elements into the main array
	if (leftArrayIndex < leftArraySize) {
		for (int index = leftArrayIndex; index < leftArraySize; index++) {
			sortedArray[mainArrayIndex] = leftArray[index];
			mainArrayIndex++;
		}
	}

	if (rightArrayIndex < rightArraySize) {
		for (int index = rightArrayIndex; index < rightArraySize; index++) {
			sortedArray[mainArrayIndex] = rightArray[index];
			mainArrayIndex++;
		}
	}
	// freeing up memory when no longer using them any more
	delete [] leftArray;
	delete [] rightArray;
}


// swap two values
void swap(double *num1, double *num2) {
	double intermediateValue = *num1;
	*num1 = *num2;
	*num2 = intermediateValue;
}

// display array to screen
void printArray(double *array, int size) {
	for (int index = 0; index < size; index++) {
		cout << array[index] << " ";
	}
}

File: SortingAlgorithms/Python/SortAlgorithms/MergeSort.py
import sys
import functools

# Modify the build-in function print (remove character a default new line
# at the end of a string)
print = functools.partial(print, end="")
sys.setrecursionlimit(1500)# allocate more memory for stack


def mergeSort(unsortedList, left, right) :
    if(left < right) :
        middle = int((left + right) / 2)
        mergeSort(unsortedList, left, middle)
        mergeSort(unsortedList, middle + 1, right)
        merge(unsortedList, left, middle, right) # should cast to right type before passing parameters (because default type is float)

 # Merge two sorted arrays into an root array
def merge(unsortedList, left, middle, right) :
    leftListSize = int(middle - left + 1)
    rightListSize = int(right - middle)

    leftList = leftListSize * [0]
    rightList = rightListSize * [0]

    for index in range(0, leftListSize) :
        leftList[index] = unsortedList[left + index]

    for index in range(0, rightListSize) :
        rightList[index] = unsortedList[middle + 1 + index]

    leftListIndex = 0
    rightListIndex = 0
    mainListIndex = left
    while ((leftListIndex < leftListSize) and (rightListIndex < rightListSize)) :
        if(leftList[leftListIndex] < rightList[rightListIndex]) :
            unsortedList[mainListIndex] = leftList[leftListIndex]
            leftListIndex += 1
        else :
            unsortedList[mainListIndex] = rightList[rightListIndex]
            rightListIndex += 1
        mainListIndex += 1

    if(leftListIndex < leftListSize) :
        for index in range(leftListIndex, leftListSize) :
            unsortedList[mainListIndex] = leftList[index]
            mainListIndex += 1

    if(rightListIndex < rightListSize) :
        for index in range(rightListIndex, rightListSize) :
            unsortedList[mainListIndex] = rightList[index]
            mainListIndex += 1

    del leftList[:]
    del rightList[:]


# Display the list on the screen
def printList(list) :
    size = len(list)
    for index in range(0, size) :
        print("%f " % list[index])
    print("\n")

 # Demo the algorithm
unsortedList = [-12.3, 6.5, 27.8, -10, 30.85]
printList(unsortedList)
mergeSort(unsortedList, 0, len(unsortedList) - 1)
printList(unsortedList)